Redirect is so old school (@redirect-induced crash(es) in some mod combination contexts)

https://github.com/TheWinABagel/Zenith/blob/399d3e3be84a2903f0268c0378ce1d9d6e17dbd5/src/main/java/dev/shadowsoffire/apotheosis/mixin/compat/clean_tooltips/absent/ItemStackMixin.java#L48C4-L48C187

Please use MixinExtras' [WrapOperation](https://github.com/LlamaLad7/MixinExtras/wiki/WrapOperation) instead

```
net.fabricmc.loader.impl.FormattedException: java.lang.BootstrapMethodError: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1799 failed
	at net.fabricmc.loader.impl.FormattedException.ofLocalized(FormattedException.java:63) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:482) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.16.5.jar:?]
	at org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:105) ~[NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.listen(EntryPoint.java:129) ~[NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.main(EntryPoint.java:70) ~[NewLaunch.jar:?]
Caused by: java.lang.BootstrapMethodError: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1799 failed
	at net.minecraft.class_5620.method_32211(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_156.method_654(class_156.java:436) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.method_32206(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.<clinit>(class_5620.java:33) ~[client-intermediary.jar:?]
	at net.minecraft.class_5546.<init>(class_5546.java:17) ~[client-intermediary.jar:?]
	at net.minecraft.class_2246.<clinit>(class_2246.java:477) ~[client-intermediary.jar:?]
	at net.minecraft.class_2358.method_10199(class_2358.java:326) ~[client-intermediary.jar:?]
	at net.minecraft.class_2966.method_12851(class_2966.java:54) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:166) ~[minecraft-1.20.1-client.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480) ~[fabric-loader-0.16.5.jar:?]
	... 5 more
Caused by: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1799 failed
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:427) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119) ~[fabric-loader-0.16.5.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at net.minecraft.class_5620.method_32211(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_156.method_654(class_156.java:436) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.method_32206(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.<clinit>(class_5620.java:33) ~[client-intermediary.jar:?]
	at net.minecraft.class_5546.<init>(class_5546.java:17) ~[client-intermediary.jar:?]
	at net.minecraft.class_2246.<clinit>(class_2246.java:477) ~[client-intermediary.jar:?]
	at net.minecraft.class_2358.method_10199(class_2358.java:326) ~[client-intermediary.jar:?]
	at net.minecraft.class_2966.method_12851(class_2966.java:54) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:166) ~[minecraft-1.20.1-client.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480) ~[fabric-loader-0.16.5.jar:?]
	... 5 more
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:422) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119) ~[fabric-loader-0.16.5.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at net.minecraft.class_5620.method_32211(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_156.method_654(class_156.java:436) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.method_32206(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.<clinit>(class_5620.java:33) ~[client-intermediary.jar:?]
	at net.minecraft.class_5546.<init>(class_5546.java:17) ~[client-intermediary.jar:?]
	at net.minecraft.class_2246.<clinit>(class_2246.java:477) ~[client-intermediary.jar:?]
	at net.minecraft.class_2358.method_10199(class_2358.java:326) ~[client-intermediary.jar:?]
	at net.minecraft.class_2966.method_12851(class_2966.java:54) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:166) ~[minecraft-1.20.1-client.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480) ~[fabric-loader-0.16.5.jar:?]
	... 5 more
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector zenith$enchTooltipRewrite(Ljava/util/List;Lnet/minecraft/class_2499;)V in zenith.mixins.json:compat.clean_tooltips.absent.ItemStackMixin from mod zenith failed injection check, (0/1) succeeded. Scanned 0 target(s). Using refmap zenith-refmap.json
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.postInject(InjectionInfo.java:531) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1490) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:752) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:330) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:246) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:437) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:418) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.15.3+mixin.0.8.7.jar:0.15.3+mixin.0.8.7]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:422) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.16.5.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119) ~[fabric-loader-0.16.5.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at net.minecraft.class_5620.method_32211(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_156.method_654(class_156.java:436) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.method_32206(class_5620.java:39) ~[client-intermediary.jar:?]
	at net.minecraft.class_5620.<clinit>(class_5620.java:33) ~[client-intermediary.jar:?]
	at net.minecraft.class_5546.<init>(class_5546.java:17) ~[client-intermediary.jar:?]
	at net.minecraft.class_2246.<clinit>(class_2246.java:477) ~[client-intermediary.jar:?]
	at net.minecraft.class_2358.method_10199(class_2358.java:326) ~[client-intermediary.jar:?]
	at net.minecraft.class_2966.method_12851(class_2966.java:54) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:166) ~[minecraft-1.20.1-client.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480) ~[fabric-loader-0.16.5.jar:?]
```

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/compat/clean_tooltips/absent/ItemStackMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/compat/clean_tooltips/absent/ItemStackMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.compat.clean_tooltips.absent;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
 import dev.shadowsoffire.placebo.events.PlaceboEventFactory;
 import net.minecraft.ChatFormatting;
@@ -16,7 +18,6 @@ import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Unique;
 import org.spongepowered.asm.mixin.injection.At;
-import org.spongepowered.asm.mixin.injection.Redirect;
 
 import java.util.*;
 
@@ -41,12 +42,8 @@ public abstract class ItemStackMixin {
         tooltip.add(mc);
     }
 
-    /**
-     * Rewrites the enchantment tooltip lines to include the effective level, as well as the (NBT + bonus) calculation.
-     */
-    @SuppressWarnings("deprecation")
-    @Redirect(method = "getTooltipLines", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/ItemStack;appendEnchantmentNames(Ljava/util/List;Lnet/minecraft/nbt/ListTag;)V"))
-    public void zenith$enchTooltipRewrite(List<Component> tooltip, ListTag tagEnchants) {
+    @WrapOperation(method = "getTooltipLines", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/ItemStack;appendEnchantmentNames(Ljava/util/List;Lnet/minecraft/nbt/ListTag;)V"))
+    public void zenith$tooltipAddEnchantmentLinesWrapper(List<Component> tooltip, ListTag tagEnchants, Operation<Void> original) {
         ItemStack ths = (ItemStack) (Object) this;
         Map<Enchantment, Integer> realLevels = new HashMap<>(EnchantmentHelper.getEnchantments(ths));
         List<Component> enchTooltips = new ArrayList<>();

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/compat/clean_tooltips/absent/ItemStackMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/compat/clean_tooltips/absent/ItemStackMixin.java
@@ -16,7 +16,6 @@ import net.minecraft.world.item.Items;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Unique;
 import org.spongepowered.asm.mixin.injection.At;
 
 import java.util.*;
@@ -24,8 +23,7 @@ import java.util.*;
 @Mixin(value = ItemStack.class, priority = 500)
 public abstract class ItemStackMixin {
 
-    @Unique
-    private static void appendModifiedEnchTooltip(List<Component> tooltip, Enchantment ench, int realLevel, int nbtLevel) {
+    private static void zenith$appendModifiedEnchTooltip(List<Component> tooltip, Enchantment ench, int realLevel, int nbtLevel) {
         MutableComponent mc = ench.getFullname(realLevel).copy();
         mc.getSiblings().clear();
         Component nbtLevelComp = Component.translatable("enchantment.level." + nbtLevel);
@@ -42,6 +40,9 @@ public abstract class ItemStackMixin {
         tooltip.add(mc);
     }
 
+    /**
+     * Modifies the enchantment tooltip lines to include the effective level, as well as the (NBT + bonus) calculation.
+     */
     @WrapOperation(method = "getTooltipLines", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/ItemStack;appendEnchantmentNames(Ljava/util/List;Lnet/minecraft/nbt/ListTag;)V"))
     public void zenith$tooltipAddEnchantmentLinesWrapper(List<Component> tooltip, ListTag tagEnchants, Operation<Void> original) {
         ItemStack ths = (ItemStack) (Object) this;
@@ -65,7 +66,7 @@ public abstract class ItemStackMixin {
             }
             else {
                 // Show the change vs nbt level
-                appendModifiedEnchTooltip(enchTooltips, ench, realLevel, nbtLevel);
+                zenith$appendModifiedEnchTooltip(enchTooltips, ench, realLevel, nbtLevel);
             }
         }
 
@@ -75,9 +76,7 @@ public abstract class ItemStackMixin {
         if(ths.is(Items.ENCHANTED_BOOK)) return;
         // Show the tooltip for any modified enchantments not present in NBT.
         for (Map.Entry<Enchantment, Integer> real : realLevels.entrySet()) {
-            if (real.getValue() > 0) appendModifiedEnchTooltip(tooltip, real.getKey(), real.getValue(), 0);
+            if (real.getValue() > 0) zenith$appendModifiedEnchTooltip(tooltip, real.getKey(), real.getValue(), 0);
         }
     }
-
-
 }

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/EnchantmentHelperMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/EnchantmentHelperMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.ench;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
 import dev.shadowsoffire.apotheosis.ench.table.RealEnchantmentHelper;
@@ -11,7 +13,6 @@ import net.minecraft.world.item.enchantment.EnchantmentInstance;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.Redirect;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
 
 import java.util.Collections;
@@ -49,18 +50,18 @@ public class EnchantmentHelperMixin {
         }
     }
 
-    @Redirect(method = "getAvailableEnchantmentResults", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.isTreasureOnly ()Z"))
-    private static boolean zenith_redirectTreasureOnly(Enchantment ench) {
+    @WrapOperation(method = "getAvailableEnchantmentResults", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/enchantment/Enchantment;isTreasureOnly()Z"))
+    private static boolean zenith_redirectTreasureOnly(Enchantment ench, Operation<Boolean> original) {
         return EnchHooks.isTreasureOnly(ench);
     }
 
-    @Redirect(method = "getAvailableEnchantmentResults", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.isDiscoverable ()Z"))
-    private static boolean zenith_redirectDiscoverable(Enchantment ench) {
+    @WrapOperation(method = "getAvailableEnchantmentResults", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/enchantment/Enchantment;isDiscoverable()Z"))
+    private static boolean zenith_redirectDiscoverable(Enchantment ench, Operation<Boolean> original) {
         return EnchHooks.isDiscoverable(ench);
     }
 
-    @Redirect(method = "getAvailableEnchantmentResults", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
-    private static int zenith_redirectMaxLevel(Enchantment ench) {
+    @WrapOperation(method = "getAvailableEnchantmentResults", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
+    private static int zenith_redirectMaxLevel(Enchantment ench, Operation<Boolean> original) {
         return EnchHooks.getMaxLevel(ench);
     }
 }

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/ItemMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/ItemMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.ench;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.EnchModule;
 import dev.shadowsoffire.apotheosis.ench.table.IEnchantableItem;
@@ -8,7 +10,6 @@ import net.minecraft.world.item.ItemStack;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.Redirect;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
 
 @Mixin(Item.class)
@@ -28,11 +29,9 @@ public abstract class ItemMixin implements IEnchantableItem {
         if (Apotheosis.enableEnch && EnchModule.isVanillaAnvil(stack)) cir.setReturnValue(true);
     }
 
-    @Redirect(method = "isEnchantable(Lnet/minecraft/world/item/ItemStack;)Z", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/Item;canBeDepleted()Z"))
-    private boolean zenith$ignoreDamageForEnchantable(Item ths, ItemStack stack) {
+    @WrapOperation(method = "isEnchantable(Lnet/minecraft/world/item/ItemStack;)Z", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/Item;canBeDepleted()Z"))
+    private boolean zenith$ignoreDamageForEnchantable(Item ths, Operation<Boolean> original) {
         return true;
     }
-
-
 }
 

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/anvil/AnvilMenuMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/anvil/AnvilMenuMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.ench.anvil;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
 import dev.shadowsoffire.apotheosis.util.Events;
@@ -47,8 +49,8 @@ public abstract class AnvilMenuMixin extends ItemCombinerMenu {
      * @param player The player using the anvil.
      * @param level  The negative of the cost of performing the anvil operation.
      */
-    @Redirect(method = "onTake", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/entity/player/Player;giveExperienceLevels(I)V"))
-    public void zenith$chargeOptimalLevels(Player player, int level) {
+    @WrapOperation(method = "onTake", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/entity/player/Player;giveExperienceLevels(I)V"))
+    public void zenith$chargeOptimalLevels(Player player, int level, Operation<Void> original) {
         EnchantmentUtils.chargeExperience(player, EnchantmentUtils.getTotalExperienceForLevel(-level));
     }
 

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/anvil/AnvilMenuMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/anvil/AnvilMenuMixin.java
@@ -98,8 +98,7 @@ public abstract class AnvilMenuMixin extends ItemCombinerMenu {
 
     @WrapOperation(method = "createResult", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/enchantment/Enchantment;getMaxLevel()I"))
     private int zenithModifyMaxLevel(Enchantment enchantment, Operation<Integer> original) {
-        if (!Apotheosis.enableEnch) return enchantment.getMaxLevel();
-        int i = enchantment.getMaxLevel();
-        return Math.max(EnchHooks.getMaxLevel(enchantment), Math.min(i, original.call(enchantment)));
+        if (!Apotheosis.enableEnch) return original.call(enchantment);
+        return EnchHooks.getMaxLevel(enchantment);
     }
 }

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/anvil/AnvilMenuMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/anvil/AnvilMenuMixin.java
@@ -96,10 +96,10 @@ public abstract class AnvilMenuMixin extends ItemCombinerMenu {
         return true;
     }
 
-    @Redirect(method = "createResult", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/enchantment/Enchantment;getMaxLevel()I"))
-    private int zenithModifyMaxLevel(Enchantment enchantment) {
+    @WrapOperation(method = "createResult", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/enchantment/Enchantment;getMaxLevel()I"))
+    private int zenithModifyMaxLevel(Enchantment enchantment, Operation<Integer> original) {
         if (!Apotheosis.enableEnch) return enchantment.getMaxLevel();
-        return EnchHooks.getMaxLevel(enchantment);
+        int i = enchantment.getMaxLevel();
+        return Math.max(EnchHooks.getMaxLevel(enchantment), Math.min(i, original.call(enchantment)));
     }
-
 }

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/BowItemMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/BowItemMixin.java
@@ -1,33 +1,31 @@
 package dev.shadowsoffire.apotheosis.mixin.ench.enchantment;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
+import com.llamalad7.mixinextras.sugar.Local;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.Ench;
-import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.BowItem;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
-import net.minecraft.world.item.enchantment.Enchantments;
-import net.minecraft.world.level.Level;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
-import org.spongepowered.asm.mixin.injection.Redirect;
 
 @Mixin(BowItem.class)
 public abstract class BowItemMixin {
 
-    @Redirect(method = "releaseUsing", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/EnchantmentHelper.getItemEnchantmentLevel (Lnet/minecraft/world/item/enchantment/Enchantment;Lnet/minecraft/world/item/ItemStack;)I", ordinal = 0))
-    private int endlessQuiver(Enchantment enchantment, ItemStack stack){
-        int infinityLevel = EnchantmentHelper.getItemEnchantmentLevel(Enchantments.INFINITY_ARROWS, stack);
+    @WrapOperation(method = "releaseUsing", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/EnchantmentHelper.getItemEnchantmentLevel (Lnet/minecraft/world/item/enchantment/Enchantment;Lnet/minecraft/world/item/ItemStack;)I", ordinal = 0))
+    private int endlessQuiver(Enchantment enchantment, ItemStack stack, Operation<Integer> original) {
+        int infinityLevel = original.call(enchantment, stack);
         if (!Apotheosis.enableEnch) return infinityLevel;
         int endlessLevel = EnchantmentHelper.getItemEnchantmentLevel(Ench.Enchantments.ENDLESS_QUIVER, stack);
         return endlessLevel > 0 ? endlessLevel : infinityLevel;
     }
 
-    @Redirect(method = "releaseUsing", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/ItemStack;is(Lnet/minecraft/world/item/Item;)Z"))
-    private boolean zenithEndlessQuiver(ItemStack stack, Item item, ItemStack bow, Level level, LivingEntity livingEntity, int i) {
-        return Apotheosis.enableEnch ? Ench.Enchantments.ENDLESS_QUIVER.isTrulyInfinite(stack, bow) || stack.is(Items.ARROW) : stack.is(Items.ARROW);
+    @WrapOperation(method = "releaseUsing", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/item/ItemStack;is(Lnet/minecraft/world/item/Item;)Z"))
+    private boolean zenithEndlessQuiver(ItemStack instance, Item item, Operation<Boolean> original, @Local(argsOnly = true) ItemStack stack) {
+        return Apotheosis.enableEnch ? Ench.Enchantments.ENDLESS_QUIVER.isTrulyInfinite(instance, stack) || original.call(instance, item) : original.call(instance, item);
     }
 }

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/EnchantCommandMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/EnchantCommandMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.ench.enchantment;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
 import net.minecraft.server.commands.EnchantCommand;
@@ -11,9 +13,9 @@ import org.spongepowered.asm.mixin.injection.Redirect;
 @Mixin(EnchantCommand.class)
 public abstract class EnchantCommandMixin {
 
-    @Redirect(method = "enchant", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
-    private static int zenithMax(Enchantment instance){
-        if (!Apotheosis.enableEnch) return instance.getMaxLevel();
+    @WrapOperation(method = "enchant", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
+    private static int zenithMax(Enchantment instance, Operation<Integer> original){
+        if (!Apotheosis.enableEnch) return original.call(instance);
         return EnchHooks.getMaxLevel(instance);
     }
 }

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/EnchantRandomlyFunctionMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/EnchantRandomlyFunctionMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.ench.enchantment;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.EnchModule;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
@@ -14,7 +16,6 @@ import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.ModifyVariable;
-import org.spongepowered.asm.mixin.injection.Redirect;
 import org.spongepowered.asm.mixin.injection.Slice;
 
 import java.util.List;
@@ -23,8 +24,9 @@ import java.util.stream.Collectors;
 @Mixin(value = EnchantRandomlyFunction.class, priority = 1100)// Patches vanilla loot to include possibility of config set level enchants
 public abstract class EnchantRandomlyFunctionMixin {
 
-    @Redirect(method = "enchantItem", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
-    private static int redirectLootLevel(Enchantment instance){
+    @WrapOperation(method = "enchantItem", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
+    private static int redirectLootLevel(Enchantment instance, Operation<Integer> original) {
+        if (!Apotheosis.enableEnch) return original.call(instance);
         return EnchHooks.getMaxLootLevel(instance);
     }
 

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/EnchantRandomlyFunctionMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/EnchantRandomlyFunctionMixin.java
@@ -6,14 +6,12 @@ import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.EnchModule;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
 import net.minecraft.core.registries.BuiltInRegistries;
-import net.minecraft.util.RandomSource;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.functions.EnchantRandomlyFunction;
 import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.ModifyVariable;
 import org.spongepowered.asm.mixin.injection.Slice;
@@ -30,15 +28,10 @@ public abstract class EnchantRandomlyFunctionMixin {
         return EnchHooks.getMaxLootLevel(instance);
     }
 
-    @Shadow
-    private static ItemStack enchantItem(ItemStack stack, Enchantment enchantment, RandomSource random) {
-        throw new RuntimeException("what");
-    }
-
     @ModifyVariable(method = "run",
             slice = @Slice(from = @At(value = "INVOKE", target = "net/minecraft/world/item/ItemStack.is (Lnet/minecraft/world/item/Item;)Z", shift = At.Shift.AFTER)),
             at = @At(value = "INVOKE_ASSIGN", target = "java/util/List.isEmpty ()Z"), index = 6) // blursed
-    private List redirectIsDiscoverable(List original, ItemStack stack, LootContext context) {
+    private List<Enchantment> redirectIsDiscoverable(List<Enchantment> original, ItemStack stack, LootContext context) {
         if (Apotheosis.enableEnch) {
             if (Apotheosis.enableDebug) EnchModule.LOGGER.info("Changing if enchantment is discoverable on {}", stack);
             return BuiltInRegistries.ENCHANTMENT.stream().filter(EnchHooks::isLootable).filter(enchantment -> stack.is(Items.BOOK) || enchantment.canEnchant(stack)).collect(Collectors.toList());

--- a/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/VillagerTradesMixin.java
+++ b/src/main/java/dev/shadowsoffire/apotheosis/mixin/ench/enchantment/VillagerTradesMixin.java
@@ -1,5 +1,7 @@
 package dev.shadowsoffire.apotheosis.mixin.ench.enchantment;
 
+import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
 import dev.shadowsoffire.apotheosis.Apotheosis;
 import dev.shadowsoffire.apotheosis.ench.asm.EnchHooks;
 import net.minecraft.core.registries.BuiltInRegistries;
@@ -8,7 +10,6 @@ import net.minecraft.world.item.enchantment.Enchantment;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.ModifyVariable;
-import org.spongepowered.asm.mixin.injection.Redirect;
 
 import java.util.List;
 import java.util.stream.Collectors;
@@ -16,8 +17,8 @@ import java.util.stream.Collectors;
 @Mixin(VillagerTrades.EnchantBookForEmeralds.class)
 public abstract class VillagerTradesMixin {
 
-    @Redirect(method = "getOffer", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.isTreasureOnly ()Z"))
-    private boolean zenith_redirectIsTradable(Enchantment ench) {
+    @WrapOperation(method = "getOffer", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.isTreasureOnly ()Z"))
+    private boolean zenith_redirectIsTradable(Enchantment ench, Operation<Boolean> original) {
         return EnchHooks.isTreasureOnly(ench);
     }
 
@@ -27,8 +28,9 @@ public abstract class VillagerTradesMixin {
         return BuiltInRegistries.ENCHANTMENT.stream().filter(EnchHooks::isTradeable).collect(Collectors.toList());
     }
 
-    @Redirect(method = "getOffer", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
-    private int zenith_redirectLootableLevel(Enchantment ench) {
+    @WrapOperation(method = "getOffer", at = @At(value = "INVOKE", target = "net/minecraft/world/item/enchantment/Enchantment.getMaxLevel ()I"))
+    private int zenith_redirectLootableLevel(Enchantment ench, Operation<Integer> original) {
+        if (!Apotheosis.enableEnch) return original.call(ench);
         return EnchHooks.getMaxLootLevel(ench);
     }
 }

